pkg/reconciler/topic/topic.go

```diff
@@ -0,0 +1,418 @@
+
+		case pubsub.OpsJobCompleteSuccessful:
+			topic.Status.MarkTopicReady()
+
+		case pubsub.OpsJobCreateFailed, pubsub.OpsJobCompleteFailed:
+			logger.Error("Failed to verify topic.",
+				zap.Any("propagationPolicy", topic.Spec.PropagationPolicy),
+				zap.Any("state", state),
+				zap.Error(err))
+
+			msg := "unknown"
+			if err != nil {
+				msg = err.Error()
+			}
+			topic.Status.MarkNoTopic(
+				"VerifyFailed",
+				"Failed to verify Topic: %q",
+				msg)
+			return err
+		}
+	default:
+		logger.Error("Unknown propagation policy.", zap.Any("propagationPolicy", topic.Spec.PropagationPolicy))
+		return nil
+	}
+
+	publisher, err := c.createOrUpdatePublisher(ctx, topic)
+	if err != nil {
+		logger.Error("Unable to create the publisher", zap.Error(err))
+		return err
+	}
+	topic.Status.PropagatePublisherStatus(publisher.Status.GetCondition(apis.ConditionReady))
+
+	if publisher.Status.IsReady() {
+		topic.Status.SetAddress(publisher.Status.Address.URL)
+	}
+	return nil
+}
+
+func (c *Reconciler) updateStatus(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, error) {
+	topic, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(topic.Status, desired.Status) {
+		return topic, nil
+	}
+	becomesReady := desired.Status.IsReady() && !topic.Status.IsReady()
+	// Don't modify the informers copy.
+	existing := topic.DeepCopy()
+	existing.Status = desired.Status
+
+	ch, err := c.RunClientSet.PubsubV1alpha1().Topics(desired.Namespace).UpdateStatus(existing)
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Topic %q became ready after %v", topic.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Topic", topic.Namespace, topic.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Topic, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) updateFinalizers(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, bool, error) {
+	source, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Don't modify the informers copy.
+	existing := source.DeepCopy()
+
+	var finalizers []string
+
+	// If there's nothing to update, just return.
+	exisitingFinalizers := sets.NewString(existing.Finalizers...)
+	desiredFinalizers := sets.NewString(desired.Finalizers...)
+
+	if desiredFinalizers.Has(finalizerName) {
+		if exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Add the finalizer.
+		finalizers = append(existing.Finalizers, finalizerName)
+	} else {
+		if !exisitingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Remove the finalizer.
+		exisitingFinalizers.Delete(finalizerName)
+		finalizers = exisitingFinalizers.List()
+	}
```

Comment:
This code seems unnecessary. `updateFinalizers` is only called if we changed the finalizers during reconcile. IIUC the only thing this is protecting against is if something else changes the object while we reconciled, but we can let the patch fail and try again, right?

Reply:
@n3wscott says: This is here so this method can be generic (and replaced by a generic method) independent of whatever the reconciler is doing with finalizers.